csrc/dynamic_transform.cpp

```diff
@@ -385,6 +648,219 @@ void DynamicTransformConcretizer::concretize() {
+          empty_tv_descr.empty_axes.end(),
+          [&out_tv](size_t ax) {
+            return out_tv->getRootDomain().at(ax)->isReduction();
+          });
+    };
+
+    // Given a TensorView get a vector of its maybeRFactor maybeExpandedExtents
+    auto orig_shape = [](TensorView* out_tv) -> std::vector<Val*> {
+      const auto& rfactor =
+          TensorDomain::noReductions(out_tv->getMaybeRFactorDomain());
+      std::vector<Val*> out_shape;
+      out_shape.reserve(rfactor.size());
+      for (const auto id : rfactor) {
+        out_shape.push_back(id->getMaybeExpandedExtent());
+      }
+      return out_shape;
+    };
+
+    // Replace uses whose outputs might not be empty. Many expressions are
+    // guaranteed to have empty outputs if any of the inputs are empty; for
+    // example simple unary or binary ops. In those cases, we don't need to
+    // doctor the Fusion since they will have an empty tensor downstream which
+    // will cut off their dependence, resulting in those uses becoming dead
+    // code. For example, suppose we determined tv2 is empty, and we have the
+    // following Fusion:
+    //
+    //   auto tv4 = add(tv2, tv3);
+    //   fusion.addOutput(tv4);
+    //
+    // If we know that tv2 is empty in any dimension, then either tv3 has a
+    // matching empty dimension or it is broadcast in that dimension. Either
+    // way, the corresponding dimension in tv4 will be empty, so tv4 is an empty
+    // tensor. If we replace this expression with
+    //
+    //   auto tv4 = full(shape, zeroVal());
+    //
+    // Then the tensors tv2 and tv3 will become dead code if they have no other
+    // live uses. In this case tv4 is an output tensor, so we must keep it in
+    // the Fusion.
+    //
+    // Some special expressions can convert an empty tensor into a non-empty
+    // tensor; particularly pad, cat, and reduction ops. These ops might have
+    // non-empty outputs so in order to guarantee that all non- input or
+    // output tensors are removed, we need to replace those ops with an
+    // equivalent that does not have any empty inputs. For example
+    for (auto use : tv->uses()) {
+      // If use is a ReductionOp or WelfordOp over some empty axes, replace it
+      // with a call to full().
+      if (auto rop = dynamic_cast<ReductionOp*>(use)) {
+        auto out = maybeReplaced(rop->out()->as<TensorView>());
+        if (hasEmptyRootReductionAxis(out)) {
+          auto out_shape = orig_shape(out);
+          replaceWithFull(out, out_shape);
+        }
+      } else if (auto wop = dynamic_cast<WelfordOp*>(use)) {
+        auto avg = maybeReplaced(wop->outAvg()->as<TensorView>());
+        auto var = maybeReplaced(wop->outVar()->as<TensorView>());
+        auto N = maybeReplaced(wop->outN()->as<TensorView>());
+        if (hasEmptyRootReductionAxis(avg)) {
+          auto out_shape = orig_shape(avg);
+          auto nan = IrBuilder::create<Double>(0.0 / 0.0);
```

Comment:
Can't we just use something `NAN` instead of actually doing a zero division? Is this a standard convention?

Reply:
Changed to use `std::numeric_limits<double>::quiet_NaN()`.